File: java/src/main/java/Vector.java
import java.util.Objects;

import static java.lang.Math.sqrt;

public class Vector {

    private final int x;
    private final int y;

    public Vector(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int x() {
        return this.x;
    }

    public int y() {
        return this.y;
    }

    public Vector add(Vector vector) {
        int sumOfX = x() + vector.x();
        int sumOfY = y() + vector.y();

        return new Vector(sumOfX, sumOfY);
    }

    public Vector subtract(Vector vector) {
        int differenceInX = x() - vector.x();
        int differenceInY = y() - vector.y();

        return new Vector(differenceInX, differenceInY);
    }

    public Vector multiplyBy(int multiplier) {
        return new Vector(x() * multiplier, y() * multiplier);
    }

    public int dotProductWith(Vector anotherVector) {
        int xMultiplied = x() * anotherVector.x();
        int yMultiplied = y() * anotherVector.y();

        return xMultiplied + yMultiplied;
    }

    public int sumOfSquares() {
        return dotProductWith(this);
    }

    public double magnitude() {
        return sqrt(sumOfSquares());
    }

    public double distanceFrom(Vector anotherVector) {
        return subtract(anotherVector).magnitude();
    }

    public int squaredDistanceFrom(Vector anotherVector) {
        return subtract(anotherVector).sumOfSquares();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector vector = (Vector) o;
        return x == vector.x &&
                y == vector.y;
    }

    @Override
    public int hashCode() {

        return Objects.hash(x, y);
    }
}

File: csharp/Vector Frankenstein/VectorFrankenstein/Vector.cs
using System;

namespace VectorFrankenstein
{
    public class Vector : IEquatable<Vector>
    {
        private readonly int x;
        private readonly int y;

        public Vector(int x, int y)
        {
            this.x = x;
            this.y = y;
        }

        public int X()
        {
            return this.x;
        }

        public int Y()
        {
            return this.y;
        }

        public Vector add(Vector vector)
        {
            int sumOfX = X() + vector.X();
            int sumOfY = Y() + vector.Y();

            return new Vector(sumOfX, sumOfY);
        }

        public Vector subtract(Vector vector)
        {
            int differenceInX = X() - vector.X();
            int differenceInY = Y() - vector.Y();

            return new Vector(differenceInX, differenceInY);
        }

        public Vector multiplyBy(int multiplier)
        {
            return new Vector(X() * multiplier, Y() * multiplier);
        }

        public int dotProductWith(Vector anotherVector)
        {
            int xMultiplied = X() * anotherVector.X();
            int yMultiplied = Y() * anotherVector.Y();

            return xMultiplied + yMultiplied;
        }

        public int sumOfSquares()
        {
            return dotProductWith(this);
        }

        public double magnitude()
        {
            return Math.Sqrt(sumOfSquares());
        }

        public double distanceFrom(Vector anotherVector)
        {
            return subtract(anotherVector).magnitude();
        }

        public int squaredDistanceFrom(Vector anotherVector)
        {
            return subtract(anotherVector).sumOfSquares();
        }

        public bool Equals(Vector other)
        {
            if (this == other)
            {
                return true;
            }

            if (other == null)
            {
                return false;
            }

            return this.X() == other.X()
                   && this.Y() == other.Y();
        }
    }
}
